rts/motoko-rts/src/idl.rs

```diff
@@ -486,3 +512,310 @@ unsafe extern "C" fn skip_fields(tb: *mut Buf, buf: *mut Buf, typtbl: *mut *mut
         *n -= 1;
     }
 }
+
+unsafe fn is_null_opt_reserved(typtbl: *mut *mut u8, end: *mut u8, t: i32) -> bool {
+    if is_primitive_type(t) {
+        return t == IDL_PRIM_null || t == IDL_PRIM_reserved;
+    }
+
+    // unfold t
+    let mut t = t;
+
+    let mut tb = Buf {
+        ptr: *typtbl.add(t as usize),
+        end: end,
+    };
+
+    t = sleb128_decode(&mut tb);
+
+    return t == IDL_CON_opt;
+}
+
+// https://github.com/dfinity/candid/blob/master/rust/candid/src/types/subtype.rs#L10
+// https://github.com/dfinity/candid/blob/20b84d1c1515e2c1db353ebe02b738486f835466/spec/Candid.md
+unsafe fn sub(
+    rel: &BitRel,
+    p: bool,
+    typtbl1: *mut *mut u8,
+    typtbl2: *mut *mut u8,
+    end1: *mut u8,
+    end2: *mut u8,
+    t1: i32,
+    t2: i32,
+) -> bool {
+    if t1 >= 0 && t2 >= 0 {
+        let t1 = t1 as u32;
+        let t2 = t2 as u32;
+        if rel.get(p, t1, t2) {
+            // cached: succeed!
+            return true;
+        };
+        // cache and continue
+        rel.set(p, t1, t2);
```

Comment:
That is an excellent question, and leads me to question whether I can even share the cache across calls to subtyping as in the follow on PR I'm now working on.

IIRC, our Motoko subtyping algorithm does not reset the cache, but perhaps that's because the caches isn't shared.

I wonder if I actually need two bits per pair of types and polarity. One to mark the goal as being visited, the second to cache the result.

I think this implementation is probably ok, because subtyping just fails and discards the local cache.

@nomeata, @rossberg thoughts?

Reply:
If it's a local cache, it's certainly okay. In the Rust implementation, we remove the cache when the check fails: https://github.com/dfinity/candid/blob/master/rust/candid/src/types/subtype.rs#L27. Because we are sharing the memo table for the whole deserialization.

Reply:
I've been thinking about this all day and I wonder if that actually works as intended/ Gamma is more of a visited set than a memo table of subtype results, right? I mean, we only remember positive outcomes, not negative ones, so expensive, but failing, tests will always be recomputed. 

For my follow on PR, I'm actually thinking of maintaining a separate memo table of yes/no/dunno results, consulted before entering the main subtype test and updated on exit with yes or no.  Unless there's some better way...

Reply:
Visited set is a subset of memo table: any entry in the memo table (yes/no/dunno) means it's visited. So memo table contains more information than Gamma. You are right that the Rust implementation only remembers the positive result, we probably need to use yes/no/dunno variant for the memo table.

Reply:
#3171 now remembers both positive and negative results, add the cost of using a separate visited bit per pair of types.